tests/apps/conftest.py

```diff
@@ -36,55 +36,92 @@ def list_apps_to_test(app_dir) -> List[AppInfo]:
 
     'apps/nanos#btc#1.5#5b6693b8.elf'
     """
-    # name example: nanos#btc#1.5#5b6693b8.elf
-    app_regexp = re.compile(
-        r"^(nanos|nanox|blue)#([^#]+)#([^#][\d\w\-.]+)#([a-f0-9]{8})\.elf$"
-    )
     all_apps = []
     for filename in os.listdir(app_dir):
         if "#" not in filename:
             continue
-        matching = re.match(app_regexp, filename)
-        if not matching:
+        path = os.path.join(app_dir, filename)
+        info = app_info_from_path(path)
+        if not info:
             pytest.fail(
                 f"An unexpected file was found in apps/, with a # but not matching the pattern: {filename!r}"
             )
-        assert len(matching.groups()) == 4
-
-        all_apps.append(
-            AppInfo(
-                filepath=os.path.join(app_dir, filename),
-                device=matching.group(1),
-                name=matching.group(2),
-                version=matching.group(3),
-                hash=matching.group(4),
-            )
-        )
+            continue
+        all_apps.append(info)
     return all_apps
 
 
 @pytest.fixture(scope="function")
-def app(request):
-    _app = APDUClient(request.param)
-    yield _app
-    _app.stop()
+def app(request, client):
+    return app_info_from_path(client.app)
 
 
-def pytest_generate_tests(metafunc):
+def get_apps(name):
     # retrieve the list of apps in the ../apps directory
     app_dir = os.path.join(SCRIPT_DIR, os.pardir, os.pardir, "apps")
     apps = list_apps_to_test(app_dir)
-    if not hasattr(metafunc.cls, "app_names") or not isinstance(
-        metafunc.cls.app_names, list
-    ):
-        pytest.fail(
-            f"The TestClass {metafunc.cls} does not have a correct 'app_names' attribute.\n"
-            "The 'app_names' attribute must contain a list of app names that will be"
-            "tested by this test."
-        )
-    apps = [app for app in apps if app.name in metafunc.cls.app_names]
-
-    # if a test function has an app parameter, give the list of app
-    if "app" in metafunc.fixturenames:
-        # test are run on each app
-        metafunc.parametrize("app", apps, indirect=True, scope="function")
+    apps = [app for app in apps if app.name == name]
+    return apps
+
+
+def default_btc_app():
+    app_dir = os.path.join(SCRIPT_DIR, os.pardir, os.pardir, "apps")
+    filepath = os.path.realpath(os.path.join(app_dir, "btc.elf"))
+    apps = get_apps("btc")
+    return [app for app in apps if os.path.realpath(app.filepath) == filepath]
+
+
+def idfn(app):
+    return os.path.basename(app.filepath)
```

Comment:
What is `idfn`? A function docstring pointing to pytest's documentation or anything useful to understand what it is about (Is `basename` an implementation detail or does this function really needs to return the name of a file? Is the output of the function displayed to the user in some circumstances, which requires its return value to be quite meaningful?)

Reply:
From the [documentation](https://docs.pytest.org/en/6.2.x/example/parametrize.html#different-options-for-test-ids) about the pytest fixtures `ids` parameter:

> pytest will build a string that is the test ID for each set of values in a parametrized test. These IDs can be used with -k to select specific cases to run, and they will also identify the specific case when one is failing. Running pytest with --collect-only will show the generated IDs.

```console
$ python3 -m pytest tests/apps/ -v
...            
tests/apps/test_btc.py::test_btc_get_version[nanos#btc#1.5#00000000.elf] PASSED                        [  4%]
tests/apps/test_btc.py::test_btc_get_public_key_with_user_approval[nanos#btc#1.5#00000000.elf] SKIPPED [  8%]
tests/apps/test_btc.py::test_btc_automation[nanos#btc#1.5#00000000.elf] SKIPPED                        [ 12%]
tests/apps/test_btc.py::test_btc_get_version[nanos#btc#1.6#6bd0a5f8.elf] PASSED                        [ 16%]
tests/apps/test_btc.py::test_btc_get_public_key_with_user_approval[nanos#btc#1.6#6bd0a5f8.elf] PASSED  [ 20%]
```

This is much useful than (previously):

```console
$ python3 -m pytest tests/apps/ -v
...
tests/apps/test_btc.py::test_btc_get_version[client_btc0] PASSED                        [  4%]
tests/apps/test_btc.py::test_btc_get_public_key_with_user_approval[client_btc0] SKIPPED [  8%]
tests/apps/test_btc.py::test_btc_automation[client_btc0] SKIPPED                        [ 12%]
tests/apps/test_btc.py::test_btc_get_version[client_btc1] PASSED                        [ 16%]
tests/apps/test_btc.py::test_btc_get_public_key_with_user_approval[client_btc1] PASSED  [ 20%]
tests/apps/test_btc.py::test_btc_automation[client_btc1] PASSED                         [ 24%]
```